### DISCUSSION TITLE
Refresh input devices when new device is plugin

### DISCUSSION THREAD

**Discussion by elmcapp** [djipco/webmidi#257] [Q&A]
Is there a way to get a list of new midi devices when a new device is plugin. Currently I can only get a list of new midi devices when I run the following function. However I like to know if there is a listener or a way allow new midi device to show without running the function every time a new device is connect.  something like a hot load

```
// Inputs
  WebMidi.inputs.forEach(input => console.log(input.manufacturer, input.name));
```

**Comment by djipco** [comment]
To do that, you can listen for the `connected` event of the `WebMidi` object. It will be triggered whenever a new device (input or output) is connected:

```js
WebMidi.addListener("connected", e => {
    console.log(e.port.name);
});
```
There are other such [events](https://webmidijs.org/api/classes/WebMidi#events) that might be of interest to you such as `portschanged`, `disabled`, `disconnected`, etc.

P.S. I moved the issue to a conversation since it isn't a problem with the library.

**Reply by elmcapp** [reply]
I tried
 WebMidi.addListener("connected", (e) => {
    console.log(e.port.name);
  });
  
  it will give me connected devices however if I connect a new device it does not update unless I refresh the script
I also tried portChange and it does not update. To get a refresh list you have to refresh the script

**Comment by djipco** [comment]
This should work. I know it works for me in the browser. What environment are you in?

**Reply by elmcapp** [reply]
node js 17.

**Reply by elmcapp** [reply]
Use it with node and react js

**Comment by jdlee022** [comment]
I'm also having trouble getting this event to fire using the below code:
`WebMidi.addListener("connected", e => {
    console.log(e.port.name);
});`

I was assuming that simply plugging in my midi controller and turning it on would get it to trigger, but maybe i misunderstand?

**Comment by djipco** [comment]
@jdlee022 I can replicate this problem in Node. It looks like the `jzz` module (which is used to emulate the Web MIDI API in the Node.js environment) is not properly firing `statechange` events. Let me look into it and get back to you.

**Reply by djipco** [reply]
I have opened [an issue](https://github.com/jazz-soft/JZZ/issues/61) about this with the `jzz` module team. Let's see what they say...

**Comment by djipco** [comment]
I have created issue #271 to follow-up on this problem. The author of the `jazz-midi` module has [confirmed](https://github.com/jazz-soft/jazz-midi/issues/8#issuecomment-1121826276) the problem. WEBMIDI.js relies on the `jazz-midi` module for Web MIDI API support on the Node.js platform. Basically, `statechange` events are not firing as they should in this context.

Thanks for @jdlee022 and @elmcapp for bringing this issue to light.

**Comment by rubendax** [comment]
Just want to report that listening for 'connected', 'disconnected', and 'portschanged' seems to be working fine now in Node / Electron with webmidi v3.0.21. Plugging  in / unplugging a USB MIDI device emits the event as expected for me.

**Reply by djipco** [reply]
Hmm... interesting. As far as I can tell, the `jzz` module has not been updated yet (see this [issue](https://github.com/jazz-soft/jazz-midi/issues/8)). Maybe an update to Node actually made the `jzz` module work properly?! This is odd...

**Reply by djipco** [reply]
I took some time to look at this once more. It still does not work on my end in Node.js and I'm suspecting the reason it works for you is that Electron uses the browser to execute WEBMIDI.js and not Node. 

In version 3.0.25 (just released it), there is a new `WebMidi.flavour` property that identifies which version of the library is running (`cjs`, `esm` or `iife`). If you had time to install v3.0.25 and print out the value of `WebMidi.flavour`, it could help me identify if the problem is indeed with the `jzz` module or not.

It's not always clear what happens under the hood in dual-engine environments such as Electron or NW.js.

**Reply by rubendax** [reply]
I think you are correct - WEBMIDI.js is being executed in the browser by Electron. `WebMidi.flavour` prints `iife` in my setup. Sorry for any confusion that I may have caused.

**Reply by djipco** [reply]
Thanks for getting back to me so fast! If it prints `iife`, it definitely means that the browser engine is running it. So, it all makes sense now. By the way, you have nothing to be sorry about. Your help is most welcome. Cheers!
